Realm/Realm/Sync/MongoClient.cs

```diff
@@ -44,12 +47,41 @@ public class MongoClient
         /// <value>The name of the remote MongoDB service.</value>
         public string ServiceName { get; }
 
+        private static readonly ConcurrentDictionary<Type, ObjectSchema> SchemaByName = new();
+
         internal MongoClient(User user, string serviceName)
         {
             User = user;
             ServiceName = serviceName;
         }
 
+        /// <summary>
+        /// Gets a collection of documents from MongoDB that can be deserialized in Realm objects.
+        /// </summary>
+        /// <remarks>
+        /// This method is only supported for source-generated classes - i.e. ones that inherit from <see cref="IRealmObject"/>
+        /// rather than <see cref="RealmObject"/>.
+        /// The collection and database name are automatically derived from the Realm object class.
+        /// </remarks>
+        /// <typeparam name="TRealmObject">The Realm object type that matches the shape of the documents in the collection.</typeparam>
+        /// <returns>A <see cref="Collection{TRealmObject}"/> instance that exposes an API for CRUD operations on its contents.</returns>
+        public Collection<TRealmObject> GetCollection<TRealmObject>()
+            where TRealmObject : class, IRealmObjectBase
+        {
+            if (!SchemaByName.TryGetValue(typeof(TRealmObject), out var schema))
+            {
+                throw new NotSupportedException("This method is only supported for source-generated classes - i.e. ones that inherit from IRealmObject rather than RealmObject.");
+            }
+
+            return new Collection<TRealmObject>(this, schema.Name);
+        }
+
+        //TODO This has been done to avoid the reflection in the GetCollection method. Need to look at this again
+        public static void RegisterSchema(Type type, ObjectSchema schema)
```

Comment:
Yes, I'm not convinced by this either, there's another comment about it. One difference is that here we get only the source generated classes, and ignore the woven ones that don't have the serializer. 
Apart from this difference, in the other comment I was wondering if we could we use something similar to create the default schema for realm and avoid the woven initializer. I think we discussed about this at some point but I don't remember what we said about it.

Reply:
One drawback of doing it in the static class constructor is that this is only invoked when the class is referenced. It is fine here since passing the type as a generic argument will invoke the static ctor, but it wouldn't be fine for the Realm schema, since where we open the Realm, we rarely enumerate all the types that can go in there.

We could generate a module initializer, but that'll only work for C# 9 projects, whereas the woven initializer works for all projects, since it's a .net feature that has existed for a very long time.

Reply:
Makes sense.

Regarding our current issue here. We could use `RealmSchema.Default`, but:
- As this contains all types, we'd need to manually check if they are source generated (checking if they implement `IRealmObjectBase`)
- We need to add methods to the `RealmSchema` to get `ObjectSchema` given a type. Not a biggie, but we don't have such method at the moment
- It won't work for classes with `Explicit` attribute. I suppose it's debatable if this is an issue or not. 

What do you think?

Reply:
Hm... a bit annoying with the `[Explicit]` - I am split between keeping the current approach and just doing the reflection - considering it's an http call, I am not particularly worried about the cost of reflection.

Reply:
I think you're right, let's keep it with reflection for now, it's the cleanest solution.